src/VCEL.Core/Expression/Func/DefaultFunctions.cs

```diff
@@ -49,6 +49,45 @@ public DefaultFunctions()
 
             Register("now", _ => DateTime.Now, TemporalDependency.Now);
             Register("today", _ => DateTime.Today, TemporalDependency.Today);
+
+            RegisterEnsureOneArg("tolower", arg => arg.ToString().ToLower());
+            RegisterEnsureOneArg("toupper", arg => arg.ToString().ToUpper());
+
+            Register("substring", args => Substring(args));
+            RegisterEnsureTwoArgs("split", (arg1, arg2) => Split(arg1.ToString(), arg2.ToString()));
+            RegisterEnsureThreeArgs("replace", (arg1, arg2, arg3) => Replace(arg1.ToString(), arg2.ToString(), arg3.ToString()));
+        }
+
+        private string Substring(object[] args)
+        {
+            switch (args.Length)
+            {
+                case 2:
+                    {
+                        var sourStr = args[0].ToString();
+                        var startIndex = int.Parse(args[1].ToString());
```

Comment:
TryParse

Reply:
We do this on purpose, whoever doing the evaluation on expression should handle the exceptions.
Otherwise the problem is hidden inside VCEL